### PR TITLE
Bullet k-DOP and occlusion culling test.

### PR DESCRIPTION
**Pull request by jsjolund**
Occlusion culling is a camera frustum culling technique which calculates not only whether an object is inside the frustum, but also if it is visible from the point of view of the camera.

k-DOP culling (k-Discrete Oriented Polytope) is a more conventional frustum culling, which uses the broadphase bounding volume tree to accelerate the culling. In this way, not all objects in the world need to be checked whether or not they are inside the frustum.

Demo: http://youtu.be/c1ncEbdDRvo

### CODE REVIEW

**Comment by davebaol** [comment]
Awesome demo :+1:

File: tests/gdx-tests/src/com/badlogic/gdx/tests/bullet/OcclusionBuffer.java [hunk 1]
```diff
@@ -0,0 +1,445 @@
...
+/** Software rasterizer used for depth rendering and testing of bounding box triangles. Stores depth values inside a
+ * {@link FloatBuffer}. CPU rendering is used in order to avoid the frequent GPU to CPU synchronization which would be needed if
+ * hardware rendering were to be used for occlusion culling queries.
+ * <p>
+ * Based on the algorithm from the Bullet CDTestFramework, BulletSAPCompleteBoxPruningTest.cpp, written by Erwin Coumans.
+ *
+ * @author jsjolund */
+public class OcclusionBuffer implements Disposable {
+	/** Integer 3D vector (incomplete) used internally for calculations. */
+	private static class IntVector3 {
```
Comment (xoppa) [inline tests/gdx-tests/src/com/badlogic/gdx/tests/bullet/OcclusionBuffer.java]: https://github.com/libgdx/libgdx/blob/master/gdx/src/com/badlogic/gdx/math/GridPoint3.java

File: tests/gdx-tests/src/com/badlogic/gdx/tests/bullet/OcclusionBuffer.java [hunk 2]
```diff
@@ -0,0 +1,445 @@
...
+					return false;
+				case QUERY:
+					return (newDepth >= oldDepth);
+			}
+			return false;
+		}
+	}
+
+	/** 4D vector (incomplete) used internally for calculations. */
+	private static class Vector4 extends Vector3 {
```
Comment (xoppa) [inline tests/gdx-tests/src/com/badlogic/gdx/tests/bullet/OcclusionBuffer.java]: What's wrong with quaternion?

File: tests/gdx-tests/src/com/badlogic/gdx/tests/bullet/OcclusionBuffer.java [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+		}
+
+		public Vector4 add (final Vector4 other) {
+			super.add(other);
+			this.w += other.w;
+			return this;
+		}
+
+		@Override
+		public Vector4 mul (final Matrix4 matrix) {
```
Comment (xoppa) [inline tests/gdx-tests/src/com/badlogic/gdx/tests/bullet/OcclusionBuffer.java]: You aren't using the w component here

**Comment by xoppa** [comment]
What's the performance of this on an typical Android device?

File: tests/gdx-tests/src/com/badlogic/gdx/tests/bullet/OcclusionBuffer.java (hunk 3 above)
Comment (jsjolund) [inline tests/gdx-tests/src/com/badlogic/gdx/tests/bullet/OcclusionBuffer.java]: Thanks for the input! I will fix these issues.

**Comment by jsjolund** [comment]
I have not tried this on Android, only desktop. Will look into it.

**Comment by xoppa** [comment]
Ok. Perhaps, in the mean time, you can split this up in the change that only involves the wrapper? So we can already pull that in and it makes reviewing it bit easier.

**Comment by jsjolund** [comment]
Ok, I split out the wrapper changes. Will fix the other issues now.

**Pull request closed without merging**